pandas/tests/arrays/test_integer.py

```diff
@@ -680,6 +680,13 @@ def test_astype_str(self):
         tm.assert_numpy_array_equal(a.astype(str), expected)
         tm.assert_numpy_array_equal(a.astype("str"), expected)
 
+    def test_astype_boolean(self):
+        # https://github.com/pandas-dev/pandas/issues/31102
+        a = pd.array([1, 0, None], dtype="Int64")
```

Comment:
I've deliberately tested only 0 and 1 here. We should decide on a behavior for other values. Are all non-zero, non-NA values allowed and considered True?

Right now, BooleanArray._from_sequence doesn't allow other values

```python
In [2]: pd.array([1, 2], dtype='boolean')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-2a7683b5ab7f> in <module>
----> 1 pd.array([1, 2], dtype='boolean')

~/sandbox/pandas/pandas/core/construction.py in array(data, dtype, copy)
    282     if is_extension_array_dtype(dtype):
    283         cls = cast(ExtensionDtype, dtype).construct_array_type()
--> 284         return cls._from_sequence(data, dtype=dtype, copy=copy)
    285
    286     if dtype is None:

~/sandbox/pandas/pandas/core/arrays/boolean.py in _from_sequence(cls, scalars, dtype, copy)
    284         if dtype:
    285             assert dtype == "boolean"
--> 286         values, mask = coerce_to_array(scalars, copy=copy)
    287         return BooleanArray(values, mask)
    288

~/sandbox/pandas/pandas/core/arrays/boolean.py in coerce_to_array(values, mask, copy)
    169                 == values_object[~mask_values].astype(float)
    170             ):
--> 171                 raise TypeError("Need to pass bool-like values")
    172
    173     if mask is None and mask_values is None:

TypeError: Need to pass bool-like values
```

Reply:
I would think want to follow Python semantics (?) but probably solved separate from this

Reply:
I think there should be *some* way to go from "potentially not only 0's and 1's" to boolean, so I am fine here to follow Python and Numpy behavior for conversion (`bool(..)`, numpy's `astype`).

The question then is: should we follow this in the constructor, or are we fine with being more strict there? (I am personally fine with being more strict)

Reply:
OK, added a test for non 0 / 1 values.